File: search_algo.py
#LINEAR SEARCH
arr = [1, 3, 4, 8 ,11, 20]
n = len(arr)

x = 8
pos = -1


for i in range(n):
    if(arr[i] == x):
        pos = i
        break
if(pos == -1):
    print(f"{x} is not found in the list.")
else:
    print(f"the position of {x} is : {pos}")

#BINARY SEARCH
#iterative solution
def binarySearch(list, l, r, x):
    while(l <=r):
        mid = (l+r) / 2
        
        #Checks if x is present at mid
        if (arr[mid] == x):
            return mid
        
        # if x greater, ignore left half
        if (arr[mid] < x):
            l = mid + 1
        
        #if x is smaller, ignore right half
        else:
            r = mid - 1
    #if we've reached here, all the elements have been checked but
    #there was no match.
    return -1

#recursive solution
def binarySearch_rec(list, l, r, x):
    if (r >= l):
        mid = (r + l) / 2
        
        if (arr[mid] == x):
            return mid
        
        #if element is smaller than mid, then
        #it can only be present in left subarray
        if (arr[mid] > x):
            return binarySearch_rec(arr, l, mid -1, x)
        
        #else the element can only be present
        #in right subarray
        return binarySearch_rec(arr, mid + 1, r, x)
    
    #Just like the iterative solution, this is where we've
    #searched all the elements but no match was found.
    return -1

#library function for binary search
from bisect import bisect_left 
  
def BinarySearch(a, x): 
    i = bisect_left(a, x) 
    if i != len(a) and a[i] == x: 
        return i 
    else: 
        return -1
  
a  = [1, 2, 4, 4, 8] 
x = int(4) 
res = BinarySearch(a, x) 
if res == -1: 
    print(x, "is absent") 
else: 
    print("First occurrence of", x, "is present at", res) 

File: search_algo.cpp
#include<bits/stdc++.h> 
using namespace std;
int binarySearch(int arr[], int l, int r, int x)
{
    while (l <= r) {
        int m = (l+r)/2;
        
        // Check if x is present at mid
        if (arr[m] == x)
            return m;
        
        // IF x greater, ignore left half
        if(arr[m] < x)
            l = m + 1;
        
        // If x is smaller, ignore right half
        else
            r = m -1;
            
    }
    
    // if we reach here, then element
    // is not present
    return -1;
}

//Recursive Implementation of Binary search
int binarySearch_rec(int arr[], int l, int r, int x)
{
    if (r >=l) {
        int mid = (r + l)/2;
        
        // Checks if the element is in the middle
        if (arr[mid] == x)
            return mid;
        
        //Checks if element is smaller than the element at mid
        // it can only be present in left subarray
        if (arr[mid] > x)
            return binarySearch_rec(arr, l, mid - 1, x);
        
        // Else the element can only be present in the right subarray
        return binarySearch_rec(arr, mid + 1, r, x);
    }
    
    // We reach here when element is not present in the sorted array 
    return -1;
}

int main() {
    int arr[] = {1, 2, 3, 4, 8, 11, 20};
    int n = 6; // size of arr;
    
    int x = 8;
    int pos  = -1;
    
    // Linear Search
    for (int i = 0; i < n; i++){
        if(arr[i] == x){
            pos = i;
            break;
        }
    }
    
    if(pos == -1) cout << "x is not found "<<endl;
    else cout<<"the position of x is : " << pos << endl;
    
    //Binary Search
    pos = binarySearch(arr, 0, n-1, x);
    
    if(pos==-1) cout <<"x is not found " <<endl;
    else cout <<"the position of x is : " <<pos << endl;
    
    // STL predefined function
    
    vector<int>vec = {1, 3, 4, 8, 11, 20};
    n = vec.size(); // size of vec
    
    int val1 = 8;
    bool exist1 = binary_search(vec.begin(), vec.end(), val1);
    
    int val2 = 5;
    bool exist2= binary_search(vec.begin(), vec.end(), val2);
    
    cout<<val1<<" is "<<(exist1 ? "found": "not found")<<endl;
    cout<<val2<<" is "<<(exist2 ? "found": "not found")<<endl;
    return 0;
}